Runtime/Data/Operation/Cache/ValueCache.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace Zinnia.Data.Operation.Cache
+{
+    using Malimbe.BehaviourStateRequirementMethod;
+    using UnityEngine;
+    using UnityEngine.Events;
+
+    /// <summary>
+    /// Caches <see cref="TValue"/> data and invokves an appropriate event when the cache is updated.
+    /// </summary>
+    /// <typeparam name="TValue">The type of data to cache.</typeparam>
+    /// <typeparam name="TEvent">The event to invoke.</typeparam>
+    public abstract class ValueCache<TValue, TEvent> : MonoBehaviour where TEvent : UnityEvent<TValue>, new()
+    {
+        /// <summary>
+        /// Emitted when the cached data is updated and has been modified to a new value.
+        /// </summary>
+        public TEvent Modified = new TEvent();
+        /// <summary>
+        /// Emitted when the cached data is updated but the value is unmodified.
+        /// </summary>
+        public TEvent Unmodified = new TEvent();
+
+        /// <summary>
+        /// The cache of data.
+        /// </summary>
+        public TValue Data
+        {
+            get
+            {
+                return data;
+            }
+            [RequiresBehaviourState]
+            set
+            {
+                if (AreEqual(value, data))
+                {
+                    Unmodified?.Invoke(value);
+                }
+                else
+                {
+                    Modified?.Invoke(value);
+                }
+                data = value;
+            }
+        }
+        /// <summary>
+        /// The backing field for <see cref="Data"/>.
+        /// </summary>
+        private TValue data;
+
+        /// <summary>
+        /// Clears the cache by setting it to the default value.
+        /// </summary>
+        [RequiresBehaviourState]
+        public virtual void ClearCache()
```

Comment:
Can't we use `[Cleared]` on the property `Data` instead? The only difference is that the (autogenerated) method will then be called `ClearData` instead - is that the reason why the attribute isn't used?

Or I guess it's due to wanting the potential ability to override this method?

Reply:
Yeah probably could use Cleared, I'll add it and run the tests. I don't think there's any need to override the method, if you have a cache with a custom clear up then you can just write a custom method at that point i think

Reply:
Putting `[Cleared]` on the `Data` property doesn't give me a `ClearData` method.

I would have thought it would, maybe it's because of my custom setter?

Reply:
Oh yeah might be. Let's just ignore this then!